ci_tools/summarise_doccoverage.py

```diff
@@ -0,0 +1,62 @@
+""" Script to check if documentation coverage has decreased
+"""
+import argparse
+import sys
+
+parser = argparse.ArgumentParser(description='Check doc coverage change')
+parser.add_argument('base', metavar='head_cov', type=str,
+                        help='File containing the coverage of the head branch')
+parser.add_argument('compare', metavar='base_cov', type=str,
+                        help='File containing the coverage of the base branch')
+
+args = parser.parse_args()
+
+results = {}
+for branch_file in [args.base, args.compare]:
+    branch = branch_file[:-4]
+    with open(branch_file, encoding="utf-8") as f:
+        lines = f.readlines()
+
+    results[branch + '_summary'] = ''.join(lines[-3:])
+    results[branch + '_no_mod'] = set()
+    results[branch + '_no_obj'] = set()
+    lines = [l for l in lines if l.startswith('File: ') or l.startswith(' - ')]
+    n = len(lines)
+    i = 0
+    while i < n:
+        modname = lines[i].split()[1].strip('"')[:-3].replace('/','.').split(branch, 1)[1][1:]
+        i+=1
+        while i<n and lines[i].startswith(' - '):
+            if lines[i].startswith(' - No module docstring'):
+                results[branch + '_no_mod'].update([modname])
+            else:
+                objname = lines[i].split()[-1].strip('`')
+                results[branch + '_no_obj'].update(['.'.join([modname,objname])])
+            i += 1
+
+added_mod = [mod for mod in results['compare_no_mod'] if mod not in results['base_no_mod']]
+added_obj = [obj for obj in results['compare_no_obj'] if obj not in results['base_no_obj']]
+
+if len(added_mod) > 0 or len(added_obj) > 0:
+    print('Failure: Coverage has decreased!')
+    print('Base Branch Summary')
+    print(results['base_summary'])
+    print('Compare Branch Summary')
+    print(results['compare_summary'])
+    if len(added_mod) > 0:
+        print('This pull request added these modules without docstrings:')
+        for mod in added_mod:
+            print(f'\t * {mod}')
+        print()
+    if len(added_obj) > 0:
+        print('This pull request added these objects without docstrings:')
+        for obj in added_obj:
+            print(f'\t * {obj}')
+        print()
+    sys.exit(1)
+else:
+    print('Success!')
+    print('Base Branch Summary')
+    print(results['base_summary'])
+    print('Compare Branch Summary')
+    print(results['compare_summary'])
```

Comment:
Why not output to the summary file so you can have a neat markdown output of this?

Reply:
I was confused about this point. I am working under the assumption that marking a test as failing is by exiting with an error code. That's why i thought that i should output the results then exit. Is there another way of marking the test as failing?

Reply:
See [here](https://github.com/pyccel/pyccel/blob/master/.github/workflows/pyccel_lint.yml) or [here](https://github.com/pyccel/pyccel/blob/master/.github/workflows/spelling.yml) for an example.

Tests are indeed marked as failing by exiting with an error code. But if you write to the file in the $GITHUB_STEP_SUMMARY variable then it appears as a markdown summary (this happens pass or fail). See for example https://github.com/pyccel/pyccel/actions/runs/3892223663

Reply:
Can you point me at an example?

Reply:
[here](https://github.com/pyccel/pyccel/actions/runs/3928835228) is one that fails, and [another](https://github.com/pyccel/pyccel/actions/runs/3997716493) that passes

Reply:
Why does the failure start with "Success!"?

Reply:
Otherwise it looks good

Reply:
The first part is a success, it did not add any new functions without a doctring. It is the second part that checks `numpydoc` validity that fails. I changed the second failure to be a header as well.

Reply:
:+1:

Reply:
@EmilyBourne Has this conversation been resolved?